fawltydeps/traverse_project.py

```diff
@@ -22,43 +22,58 @@ def find_sources(
     and yield the corresponding *Source objects found.
     """
 
-    # Collect any directories we will need to traverse.
-    dirs_to_traverse: Dict[Path, Set[Type[Source]]] = {}
+    logger.debug("find_sources() Looking for sources under:")
+    logger.debug(f"    code:   {settings.code}")
+    logger.debug(f"    deps:   {settings.deps}")
+
+    traversal: DirectoryTraversal[Union[Type[Source], Path]] = DirectoryTraversal()
 
     for path_or_special in settings.code if CodeSource in source_types else []:
         # exceptions raised by validate_code_source() are propagated here
         validated: Optional[Source] = validate_code_source(path_or_special)
         if validated is not None:  # parse-able file given directly
+            logger.debug(f"find_sources() Found {validated}")
             yield validated
         else:  # must traverse directory
             # sanity check: convince mypy that SpecialPath is already handled
             assert isinstance(path_or_special, Path)
-            dirs_to_traverse.setdefault(path_or_special, set()).add(CodeSource)
+            traversal.add(path_or_special, CodeSource)
+            traversal.add(path_or_special, path_or_special)  # also record base dir
```

Comment:
I don't follow what this line is doing.

Reply:
We "attach" two pieces of information to the traversal of this directory:
- `CodeSource`: i.e. we want to find/create `CodeSource` objects under this directory
- `path_or_special`: This directory is the _base directory_ we want to use when creating `CodeSource` objects from python files under this directory. This "attached" information is extracted (or "detached") on line 62 below, and is then passed as the `base_dir` argument to the `validate_code_source` call. From there it is stored in the `CodeSource.base_dir` member, and finally passed to `make_isort_config()` to make `isort` correctly discern first- and third-party imports.

The reason why we need to attach this here is that with the new `DirectoryTraversal` code the _base directory_ is no longer implicitly known _while_ we traverse.

Before we had:
```python
    for dir_path, types in dirs_to_traverse.items():
        assert len(types) > 0
        for file in walk_dir(dir_path):
            ...
```
and `dir_path` was assumed to be the base directory. If `dirs_to_traverse` contained both `foo/` and `foo/bar/`, we would end up traversing `foo/bar` _twice_, first with `foo/` as base dir, and then with `foo/bar/`.

The new code fixes that issue, but at the cost of the traversal itself no longer knowing exactly which of `foo/` or `foo/bar/` would be the appropriate base dir. We fix this by "attaching" the correct base dir to the traversal object.

Reply:
That makes much more sense now! Thanks for the explanation!